react-js/react-part4-full-apps.md

```diff
@@ -0,0 +1,343 @@
+todos.map((todo, idx) => (
+              <Todo
+                key={idx}
+                todo={todo}
+                updateTodo={this.updateTodo.bind(this)}
+              />
+            ))
+```
+
+This code is a little more complicated than you might've expected. This is because it's creating a whole new copy of the array (using `.map`) rather than finding the item and changing its data.
+
+#### Rewrite `updateTodo` in your own coding style
+
+You'll learn a lot more about different ways of managing state in React, but that is a whole big topic in itself. For now, you should rewrite the `updateTodo` function in a way that makes sense for you. There are many ways to accomplish the same logic.
+
+Now test out clicking the checkbox. It should toggle correctly as expected.
+
+### Independent Practice
+
+With your daily pair, review each other's code from above and see if you can explain how each part is working, especially how you chose to rewrite the `updateTodo` function. Add some `console.log` statements to verify your understanding.
+
+Next, pair program to add a feature: sorting.
+
+#### Sorting
+
+First we need something to sort by so let's add a `createdAt` property to each item that will be a `Date` object.
+
+```js
+{
+  text: "Walk dog",
+  isCompleted: false,
+  createdAt: new Date(),
+},
+```
+
+Let's support two sort options: newest first or oldest first.
+
+Let's make newest first the default and add it to our starting state.
+
+```js
+state = {
+  sort: 'oldest',
+  todos: []
+};
+```
+
+Then based on that value, we'll sort `state.todos`.
+
+```js
+const { todos, sort } = this.state;
+if (sort === 'oldest first') {
+  todos.sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime());
+}
+```
+
+Check the browser and make sure it's sorting correctly. You may have to give the todos specific dates to test it, e.g. `new Date("2021-01-15 00:00:00-0800")`
+
+Now add another option to sort by `'newest'`.
+
+```jsx
+if (sort === 'oldest') {
+  todos.sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime());
+} else if (sort === 'newest') {
+  todos.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
+}
+```
+
+Edit the starting state and make sure the sorting works as expected.
+
+#### Sorting Toggle
+
+Now add a button that switches the sorting from one option to the other.
+
+```jsx
+<div>
+  <button
+    onClick={() =>
+      this.setState({ sort: sort === 'oldest' ? 'newest' : 'oldest' })
+    }
+  >
+    ⬇️ Sort by {sort === 'oldest' ? 'newest' : 'oldest'} first
+  </button>
+</div>
+```
+
+Test out the button to make sure it works correctly.
+
+### Challenges
+
+#### Style the Todo Items
+
+If they are complete, style them to indicate that, perhaps by graying them out or using strikethrough.
+
+#### Add Input Form
+
+To create new todos, add an input form that will create a new todo item and add it to the state so it shows up in the list.
+
+#### Add `reactstrap` npm package to your CRA project
```

Comment:
Do you think React Styling should come before or after this outline?  Does it matter?  https://github.com/Techtonica/curriculum/blob/main/react-js/styling-react.md

Reply:
Not sure where we planned for it to go in the schedule.... but I can at least link it up. I'd say a dedicated styling lesson can come later, maybe before Eventonica React.

#1160 added it during GSSOC.

Reply:
While we're on it, those example code snippets looks a bit weird, so I opened an issue to polish that lesson up: #1503